<?xml version="1.0"?>
<robot name="ur5" xmlns:controller="http://playerstage.sourceforge.net/gazebo/xmlschema/#controller" xmlns:interface="http://playerstage.sourceforge.net/gazebo/xmlschema/#interface" xmlns:sensor="http://playerstage.sourceforge.net/gazebo/xmlschema/#sensor" xmlns:xacro="http://ros.org/wiki/xacro">
  <material name="UR/Blue">
    <color rgba="0.0 0.0 0.8 1.0"/>
  </material>
  <material name="UR/Grey">
    <color rgba="0.7 0.7 0.7 1.0"/>
  </material>
  <!-- arm -->
    <!-- World link to fix basin -->
  <link name="world"/>


  <link name="base_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.6096 0.4191 0.07"/>
      </geometry>
      <material name="blue">
        <color rgba="0 0.5 1 1"/>
      </material>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.6096 0.4191 0.07"/>
      </geometry>
    </collision>

    <inertial>
      <mass value="5"/>
      <inertia ixx="0.13" ixy="0.0" ixz="0.0" iyy="0.21" iyz="0.0" izz="0.13"/>
    </inertial>
  </link>

  <!-- ========================= -->
  <!-- Wheels -->
  <!-- ========================= -->
  <!-- Right Wheel -->
  <link name="right_wheel">
    <visual>
      <origin xyz="0 0 0" rpy="1.570795 0 0"/>
      <geometry>
        <cylinder radius="0.0762" length="0.04318"/>
      </geometry>
      <material name="darkgray">
        <color rgba=".2 .2 .2 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="1.570795 0 0"/>
      <geometry>
        <cylinder radius="0.0762" length="0.04318"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.5"/>
      <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.005" iyz="0.0" izz="0.005"/>
    </inertial>
  </link>
  <joint name="joint_right_wheel" type="continuous">
    <parent link="base_link"/>
    <child link="right_wheel"/>
    <origin xyz="0.2 0.24 0" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
  </joint>

  <!-- Left Wheel -->
  <link name="left_wheel">
    <visual>
      <origin xyz="0 0 0" rpy="1.570795 0 0"/>
      <geometry>
        <cylinder radius="0.0762" length="0.04318"/>
      </geometry>
      <material name="darkgray">
        <color rgba=".2 .2 .2 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="1.570795 0 0"/>
      <geometry>
        <cylinder radius="0.0762" length="0.04318"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.5"/>
      <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.005" iyz="0.0" izz="0.005"/>
    </inertial>
  </link>
  <joint name="joint_left_wheel" type="continuous">
    <parent link="base_link"/>
    <child link="left_wheel"/>
    <origin xyz="0.2 -0.24 0" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
  </joint>

  <!-- ========================= -->
  <!-- Caster -->
  <!-- ========================= -->
  <link name="caster">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.0381"/>
      </geometry>
      <material name="white">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.0381"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.5"/>
      <inertia ixx="0.0001" ixy="0.0" ixz="0.0" iyy="0.0001" iyz="0.0" izz="0.0001"/>
    </inertial>
  </link>
  <joint name="caster_joint" type="continuous">
    <parent link="base_link"/>
    <child link="caster"/>
    <origin xyz="-0.3 0.0 -0.07" rpy="1.57 0 0"/>
  </joint>

  <!-- ========================= -->
  <!-- Pole and Plate -->
  <!-- ========================= -->
  <link name="pole">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.05" length="0.5"/>
      </geometry>
      <material name="grass">
        <color rgba="0.36 0.67 0.18 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.05" length="0.5"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="1.0"/>
      <inertia ixx="0.004375" ixy="0" ixz="0" iyy="0.004375" iyz="0" izz="0.005"/>
    </inertial>
  </link>
  <joint name="pole_joint" type="fixed">
    <parent link="base_link"/>
    <child link="pole"/>
    <origin xyz="0.225 0 0.27" rpy="0 0 3.1416"/>
  </joint>

  <!-- Rotating Plate -->
  <link name="plate">
    <visual>
      <origin xyz="0.25 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="1.0 0.1 0.02"/>
      </geometry>
      <material name="gray">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.5 0.3 0.02"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="1.0"/>
      <inertia ixx="0.01" ixy="0.0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
    </inertial>
  </link>
  <joint name="plate_joint" type="revolute">
    <parent link="pole"/>
    <child link="plate"/>
    <origin xyz="0 0 0.25" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <limit lower="-3.1416" upper="3.1416" effort="5" velocity="1"/>
  </joint>

  <!-- Plate Slider -->
  <link name="plate_slider">
    <visual>
      <origin xyz="0.5 0 0.0" rpy="0 0 0"/>
      <geometry>
        <box size="0.1 0.1 0.05"/>
      </geometry>
      <material name="red">
        <color rgba="1 0 0 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0.5 0 0.05" rpy="0 0 0"/>
      <geometry>
        <box size="0.1 0.1 0.02"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.2"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0" iyy="0.001" iyz="0" izz="0.001"/>
    </inertial>
  </link>
  <joint name="plate_slider_joint" type="prismatic">
    <parent link="plate"/>
    <child link="plate_slider"/>
    <origin xyz="0 0 0.035" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
    <limit lower="-0.5" upper="0.25" effort="5" velocity="0.1"/>
  </joint>


  <link name="base_link_1">
    <visual>
      <geometry>
        <mesh filename="package://ur5_description/meshes/ur5/visual/Base.dae"/>
      </geometry>
      <origin rpy="0.0 0.0 2.3561944875" xyz="0.0 0.0 0.003"/>
    </visual>

    <collision>
      <geometry>
        <mesh filename="package://ur5_description/meshes/ur5/collision/Base.dae"/>
      </geometry>

    </collision>
  <inertial>
    <mass value="4.0"/>
    <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
  </inertial>
  </link>
  
<joint name="base_to_ur5" type="fixed">
  <parent link="plate_slider"/>
  <child link="base_link_1"/>
  <origin xyz="0.5 0 0.025" rpy="0 0 0"/> <!-- adjust Z height as needed -->
</joint>

  <joint name="shoulder_pan_joint" type="revolute">
    <parent link="base_link_1"/>
    <child link="shoulder_link"/>
    <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.089159"/>
    <axis xyz="0 0 1"/>
    <limit effort="10.0" lower="-3.14159265" upper="3.14159265" velocity="3.14159265"/>
    <dynamics damping="10" friction="0.1"/>
  </joint>
  <link name="shoulder_link">
    <visual>
      <geometry>
        <mesh filename="package://ur5_description/meshes/ur5/visual/Shoulder.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://ur5_description/meshes/ur5/collision/Shoulder.dae"/>
      </geometry>
      <!-- <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 -0.009159"/> -->
    </collision>
    <inertial>
      <mass value="3.7"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
      <origin rpy="0.0 0.0 0.0" xyz="0.0 0.00193 -0.02561"/>
    </inertial>
  </link>
  <joint name="shoulder_lift_joint" type="revolute">
    <parent link="shoulder_link"/>
    <child link="upper_arm_link"/>
    <origin rpy="0.0 1.570796325 0.0" xyz="0.0 0.13585 0.0"/>
    <axis xyz="0 1 0"/>
    <limit effort="10.0" lower="-3.14159265" upper="3.14159265" velocity="3.14159265"/>
    <dynamics damping="10" friction="0.1"/>
  </joint>
  <link name="upper_arm_link">
    <visual>
      <geometry>
        <mesh filename="package://ur5_description/meshes/ur5/visual/UpperArm.dae"/>
      </geometry>
      <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.0"/>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://ur5_description/meshes/ur5/collision/UpperArm.dae"/>
      </geometry>
      <!--       <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.2125"/> -->
    </collision>
    <inertial>
      <mass value="8.393"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
      <origin rpy="0.0 0.0 0.0" xyz="0.0 -0.024201 0.2125"/>
    </inertial>
  </link>
  <joint name="elbow_joint" type="revolute">
    <parent link="upper_arm_link"/>
    <child link="forearm_link"/>
    <origin rpy="0.0 0.0 0.0" xyz="0.0 -0.1197 0.425"/>
    <axis xyz="0 1 0"/>
    <limit effort="10.0" lower="-3.14159265" upper="3.14159265" velocity="3.14159265"/>
    <dynamics damping="10" friction="0.1"/>
  </joint>
  <link name="forearm_link">
    <visual>
      <geometry>
        <mesh filename="package://ur5_description/meshes/ur5/visual/Forearm.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://ur5_description/meshes/ur5/collision/Forearm.dae"/>
      </geometry>
      <!--       <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.196125"/> -->
    </collision>
    <inertial>
      <mass value="2.275"/>
      <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0265 0.11993"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <joint name="wrist_1_joint" type="revolute">
    <parent link="forearm_link"/>
    <child link="wrist_1_link"/>
    <origin rpy="0.0 1.570796325 0.0" xyz="0.0 0.0 0.39225"/>
    <axis xyz="0 1 0"/>
    <limit effort="10.0" lower="-3.14159265" upper="0.7853981625" velocity="3.14159265"/>
    <dynamics damping="10" friction="0.1"/>
  </joint>
  <link name="wrist_1_link">
    <visual>
      <geometry>
        <mesh filename="package://ur5_description/meshes/ur5/visual/Wrist1.dae"/>
      </geometry>
      <origin rpy="0.0 1.570796325 0.0" xyz="0.0 0.093 0.0"/>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://ur5_description/meshes/ur5/collision/Wrist1.dae"/>
      </geometry>
      <origin rpy="0.0 1.570796325 0.0" xyz="0.0 0.093 0.0"/>
    </collision>
    <inertial>
      <mass value="1.219"/>
      <origin rpy="0.0 0.0 0.0" xyz="0.0 0.110949 0.01634"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <joint name="wrist_2_joint" type="revolute">
    <parent link="wrist_1_link"/>
    <child link="wrist_2_link"/>
    <origin rpy="0.0 0.0 0.0" xyz="0.0 0.093 0.0"/>
    <axis xyz="0 0 1"/>
    <limit effort="10.0" lower="-3.14159265" upper="3.14159265" velocity="3.14159265"/>
    <dynamics damping="10" friction="0.1"/>
  </joint>
  <link name="wrist_2_link">
    <visual>
      <geometry>
        <mesh filename="package://ur5_description/meshes/ur5/visual/Wrist2.dae"/>
      </geometry>
      <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.09465"/>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://ur5_description/meshes/ur5/collision/Wrist2.dae"/>
      </geometry>
      <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.09465"/>
    </collision>
    <inertial>
      <mass value="1.219"/>
      <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0018 0.11099"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <joint name="wrist_3_joint" type="revolute">
    <parent link="wrist_2_link"/>
    <child link="wrist_3_link"/>
    <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.09465"/>
    <axis xyz="0 1 0"/>
    <limit effort="10.0" lower="-3.14159265" upper="3.14159265" velocity="3.14159265"/>
    <dynamics damping="10" friction="0.1"/>
  </joint>
  <link name="wrist_3_link">
    <visual>
      <geometry>
        <mesh filename="package://ur5_description/meshes/ur5/visual/Wrist3.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://ur5_description/meshes/ur5/collision/Wrist3.dae"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.1879"/>
      <origin rpy="0.0 0.0 0.0" xyz="0.0 0.001159 0.0"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <joint name="ee_fixed_joint" type="fixed">
    <parent link="wrist_3_link"/>
    <child link="ee_link"/>
    <origin rpy="0.0 0.0 1.570796325" xyz="0.0 0.0823 0.0"/>
  </joint>
  <link name="ee_link"/>

  <link name="brush_link">
    <visual>
      <geometry>
        <cylinder length="0.3" radius="0.02"/>
      </geometry>
      <origin xyz="0 0 0.15" rpy="0 0 0"/> <!-- move +0.15 along Z -->
      <material name="blue">
        <color rgba="0 0 1 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.1" radius="0.02"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="0.0001" ixy="0" ixz="0" iyy="0.0001" iyz="0" izz="0.0001"/>
      <origin xyz="0 0 0.05" rpy="0 0 0"/>
    </inertial>
  </link>

    <!-- Fixed joint to attach to robot's end-effector -->
  <joint name="brush_joint" type="fixed">
    <parent link="ee_link"/> <!-- Replace ee_link with your robot's end-effector link -->
    <child link="brush_link"/>
    <origin xyz="0 0 0.0" rpy="0 1.57 0"/>
  </joint>




<!-- Joints -->

<!-- World â†’ bottom cylinder -->
<joint name="world_to_base_link" type="fixed">
  <parent link="world"/>
  <child link="base_link"/>
  <origin xyz="0 0 0.05" rpy="0 0 0"/>
</joint>


<link name="washbasin_link">
  <visual>
    <geometry>
      <mesh filename="package://ur5_description/meshes/ur5/visual/basin.STL" scale="0.001 0.001 0.001"/>
    </geometry>
    <origin xyz="0.2 0 0.8" rpy="-3.14 0 0"/> <!-- half height above ground -->
    <material name="White">
      <color rgba="0.9 0.9 0.9 1.0"/>
    </material>
  </visual>
  <collision>
    <geometry>
      <cylinder radius="0.3" length="0.1"/> <!-- simple collision -->
    </geometry>
    <origin xyz="0 0 0.05" rpy="0 0 0"/>
  </collision>
  <inertial>
    <mass value="5.0"/>
    <inertia ixx="0.1" iyy="0.1" izz="0.1" ixy="0" ixz="0" iyz="0"/>
    <origin xyz="0 0 0.05" rpy="0 0 0"/>
  </inertial>
</link>

<joint name="world_to_washbasin" type="fixed">
  <parent link="world"/>
  <child link="washbasin_link"/>
  <origin xyz="-1.3 0.15 0.05" rpy="0 0 0"/>
</joint>


<!-- Attch a counte weight -->
<link name="counterweight">
  <visual>
    <origin xyz="-0.3 0 0.05" rpy="0 0 0"/> <!-- behind slider along -X -->
    <geometry>
      <box size="0.1 0.1 0.1"/> <!-- adjust size as needed -->
    </geometry>
    <material name="gray">
      <color rgba="0.5 0.5 0.5 1"/>
    </material>
  </visual>
  <collision>
    <origin xyz="-0.3 0 0.025" rpy="0 0 0"/>
    <geometry>
      <box size="0.1 0.1 0.05"/>
    </geometry>
  </collision>
  <inertial>
    <mass value="10000.0"/>
    <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
  </inertial>
</link>

<joint name="counterweight_to_slider" type="fixed">
  <parent link="plate"/>
  <child link="counterweight"/>
  <origin xyz="0.1 0 0" rpy="0 0 0"/> <!-- relative to slider -->
</joint>


<gazebo reference="base_link_1">
    <material>Gazebo/Orange</material>
</gazebo>

<gazebo reference="shoulder_link">
    <material>Gazebo/Orange</material>
</gazebo>

<gazebo reference="upper_arm_link">
    <material>Gazebo/Orange</material>
</gazebo>

<gazebo reference="forearm_link">
    <material>Gazebo/Orange</material>
</gazebo>

<gazebo reference="wrist_1_link">
    <material>Gazebo/Red</material>
</gazebo>

<gazebo reference="wrist_2_link">
    <material>Gazebo/Orange</material>
</gazebo>

<gazebo reference="wrist_3_link">
    <material>Gazebo/Orange</material>
</gazebo>

<gazebo reference="base_link">
    <material>Gazebo/Black</material>
</gazebo>
<gazebo reference="right_wheel">
    <material>Gazebo/Black</material>
</gazebo>
<gazebo reference="left_wheel">
    <material>Gazebo/Black</material>
</gazebo>
<gazebo reference="caster">
    <material>Gazebo/Red</material>
</gazebo>
<gazebo reference="brush_link">
    <material>Gazebo/Blue</material>
</gazebo>
<gazebo reference="plate">
    <material>Gazebo/Red</material>
</gazebo>

<gazebo reference="counterweight">
    <material>Gazebo/Black</material>
</gazebo>





<!--  Addign Controllers -->

<ros2_control name="GazeboSystem" type="system">
  <hardware>
    <plugin>gazebo_ros2_control/GazeboSystem</plugin>
  </hardware>

  <!-- Shoulder Pan -->
  <joint name="shoulder_pan_joint">
      <param name="initial_position">-0.323</param>
      <command_interface name="effort"/>
      <command_interface name="velocity"/>
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
  </joint>
  <!-- Shoulder Lift -->
  <joint name="shoulder_lift_joint">
      <param name="initial_position">-0.425</param>
      <command_interface name="effort"/>
      <command_interface name="velocity"/>
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
  </joint>
  <!-- Elbow -->
  <joint name="elbow_joint">
      <param name="initial_position">0.357</param>
      <command_interface name="effort"/>
      <command_interface name="velocity"/>
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
  </joint>
  <!-- Wrist 1 -->
  <joint name="wrist_1_joint">
      <param name="initial_position">-0.849</param>
      <command_interface name="effort"/>
      <command_interface name="velocity"/>
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
  </joint>
  <!-- Wrist 2 -->
  <joint name="wrist_2_joint">
      <param name="initial_position">-1.817</param>
      <command_interface name="effort"/>
      <command_interface name="velocity"/>
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
  </joint>
  <!-- Wrist 3 -->
  <joint name="wrist_3_joint">
      <param name="initial_position">-0.493</param>
      <command_interface name="effort"/>
      <command_interface name="velocity"/>
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
  </joint>

</ros2_control>

<gazebo>
  <plugin name="gazebo_ros2_control" filename="libgazebo_ros2_control.so">
    <robot_param>robot_description</robot_param>
    <robot_param_node>robot_state_publisher</robot_param_node>
    <parameters>
      /home/vishal/UR5_AMR_CL_1/src/robotic_arms_control/robotic_arms_control/config/ur5_itc_controller.yaml
    </parameters>
  </plugin>
</gazebo>


</robot>

